Fix OAuth redirection to previous url after LOGIN

Uzay-G opened the pull request:
Fixes #6233 

The `origin` parameter for the oauth login was always empty although that was the parameter that told the omniauth where to redirect.

There were two different cases here where we need to solve the redirection. One, the user uses the signup modal to log in. Two, the user goes to a page restricted for logged in users and is redirected to the login page.

In the first one, we want to redirect the user to the page he is currently on when he is filling the signup overlay. We can access this with url in Rails with `request.origin_url`. For the second case, the user is on the "/login" page which is not where he wants to be redirected. To fix this, I edited the `require_login` filter that redirects to the login page to set a `return_to` parameter:
```ruby
def require_user
    unless current_user
      store_location
      flash[:warning] ||= I18n.t('application_controller.must_be_logged_in_to_access')
     # what I added
      redirect_to "/login?return_to=" + request.fullpath
      false
    end
end
```
So, if the user is trying to post a node at https://publiclab.org/post, the `require_login` function is called and  he will be redirected to `https://publiclab.org/login?return_to=/post`. We can then set the origin attribute for oauth to `params[:return_to]`. So now that these two problems could be solved individually, I needed to pair the two so that the origin parameter be set to the `return_to` parameter when there is one and otherwise, when the modal is being used, the origin is set to the current url. 

This is the result:
`<a href="/auth/provider?origin=<%= params[:return_to] || request.original_url %>"`

So for example, if user1 is at https://publiclab.org/about and they click the login modal, they are still on the same page and the origin will be set to `request.original_url` -> https://publiclab.org/about. But if the `params[:return_to]` is set, he will be redirected to the page he was trying to go to before logging in.

This is a demo of the working redirection for both cases:
![correct-login](https://user-images.githubusercontent.com/52892257/71781725-00b94180-2fd2-11ea-8a27-5fd0ed0f374f.gif)

I also added a test to the users_sessions controller for the oauth redirection!

Although I could not claim this task on the GCI dashboard, it has been open for a long time and I wanted to solve it because this issue was interfering with the testing of my pull request #7005 
It was really fun to investigate all the mechanisms for the issue and then documenting the fix.
I hope you're okay that I submitted the pull request without claiming it on GCI and I can retract it if you want but I could not work on tests for the other pull request because of this problem.

Thanks! :+1:

Uzay-G commented:
For now, the initial sign-in does not redirect to dashboard, but I have the changes to implement that if you want me too.

SidharthBansal commented:
Why you have changed all the tests?

SidharthBansal commented:
This leads to one more harder issues for CDSM, you can claim it after this. Just you have to write the tests in MK, SWB and/or Plots. I am not sure. We can discuss after this is merged.

Uzay-G commented:
I had to change the tests because when users were going to a restricted page, the were redirected to login + the `return_to` param. I don't really understand what you mean by your second message.

SidharthBansal commented:
> I don't really understand what you mean by your second message.

We will require a pr after merging 7113

Uzay-G commented:
> We will require a pr after merging 7113
With that purpose?

SidharthBansal commented:
> I had to change the tests because when users were going to a restricted page, the were redirected to login + the return_to param. 

Please explain with an example

SidharthBansal commented:
Please note that this pr affects Mapknitter and spectralworkbench repo, so we need to be extra carefully before merging it.
Thanks

Uzay-G commented:
Ok, here's an example before the change:
```ruby
  test 'non-registered user should not be able to see spam_revisions page' do
    UserSession.create(users(:admin))
    UserSession.find.destroy

    get :spam_revisions

    assert_equal 'You must be logged in to access this page', flash[:warning]
    assert_redirected_to '/login'
  end
```
With my changes, when the user goes to a login-restricted page, we store a `return_to` parameter in the url which is where we will redirect the user after logging in. So if he has tried to go to to `/spam/revisions/` endpoint but was blocked, his `return_to` is `/spam/revisions` and he is redirected to `/login?return_to=/spam/revisions/` and not `/login`.
So I had to change the tests to fix that error.

SidharthBansal commented:
> With that purpose?

In short, Mapknitter and SpectralWorkBench Authentication depends on Plots2 authentication System. Plots2 Auth system in indeed dependent on Oauth and omniauth.

So, we need tests for mapknitter and spectralworkbench. May be we will not require them. I have to investigate it or may be you can. Problem with your investigation: if you don't require tests then you will get 3points for research/investigation, but if there would be tests then it can be another 15 points task. 

You can do it when your pr is not reviewed.

SidharthBansal commented:
I mean, when you are in a free state.

SidharthBansal commented:
You are working on 3 repo work in this pr. So, please keep checking things. 
It's really great to see that you come so far. :-)

Uzay-G commented:
Yeah ok no problem. So I would test the oauth of SWB and mapknitter if they are affected by these changes?
Thanks! I'm learning so much about Rails it's nice! :+1:

SidharthBansal commented:
Difficult thing, but we have to:
Please check these things on localhost and provide screenshots.
Thanks

SidharthBansal commented:
by these things I mean tests in this pr.

Uzay-G commented:
Sorry, I posted a gif for the working redirection. Do you want me to provide screenshots for the oauth of Mapknitter and SWB to see if they are affected?

SidharthBansal commented:
> Sorry, I posted a gif for the working redirection. Do you want me to provide screenshots for the oauth of Mapknitter and SWB to see if they are affected?

 I am asking after rectified tests in #7113 
I am not asking for MK and SWB work now.

SidharthBansal commented:
Please claim the tasks for installation of MK and SWB and get points

Uzay-G commented:
Alright I will claim them when I am not on another task

SidharthBansal commented:
ok

On Mon, Jan 6, 2020 at 1:09 PM Uzay-G <notifications@github.com> wrote:

> Alright I will claim them when I am not on another task
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/7113?email_source=notifications&email_token=AFAAEQYFRODMKDVJQ3XBCNLQ4LN4TA5CNFSM4KC3VFC2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIEVPJI#issuecomment-571037605>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFAAEQ3X4G5BRSR4OJOBKZTQ4LN4TANCNFSM4KC3VFCQ>
> .
>

Uzay-G commented:
Sorry, I don't really understand what screenshots you want me to provide for the tests. Do you want me to follow the path on localhost of the example test I showed to make sure everything is working?

Uzay-G commented:
Here is an example of why I changed the tests:
![redirection-tests](https://user-images.githubusercontent.com/52892257/71836414-39314c00-30b4-11ea-91ed-b46f3beb6e33.gif)

As you can see, with my changes, the non-logged in user gets redirected to `/login?return_to=/login_restricted_page`. That's why I changed the tests.

SidharthBansal commented:
What happens after you login in above screenshot?

On Mon, 6 Jan 2020, 11:12 pm Uzay-G, <notifications@github.com> wrote:

> Here is an example of why I changed the tests:
> [image: redirection-tests]
> <https://user-images.githubusercontent.com/52892257/71836414-39314c00-30b4-11ea-91ed-b46f3beb6e33.gif>
>
> As you can see, with my changes, the non-logged in user gets redirected to
> /login?return_to=/login_restricted_page. That's why I changed the tests.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/7113?email_source=notifications&email_token=AFAAEQZOCCJYJSYW7MQ3PIDQ4NUR5A5CNFSM4KC3VFC2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIGF6AI#issuecomment-571236097>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFAAEQ4RBIS26WA5D4PDA5DQ4NUR5ANCNFSM4KC3VFCQ>
> .
>

Uzay-G commented:
After I login, I get redirected to `/spam/revisions` but since I am not an admin in that video, I then get redirected to the root url.

SidharthBansal commented:
What if you login as an admin?

SidharthBansal commented:
Try and provide gif/video

Uzay-G commented:
Yeah I will do it. Do you know how I can create a new admin user through oauth?

SidharthBansal commented:
Hello, I have some tasks for you while I am busy:
1. Set up MK, SWB and plots2 on local and try to connect these apps together so that MK and SWB will use OAuth of plots2. It is tough but fun. I will help you in that
2. This PR is for login or sign in or authentication( authentication means login + signup). 
Please create similar pr for sign up and change names accordingly if needed.

SidharthBansal commented:
Use existing admin account. Link them to providers.

Uzay-G commented:
Oh ok thanks!

SidharthBansal commented:
1. login as admin
2. set up google developer app
3. go to edit profile section 
4. login with google

SidharthBansal commented:
Create github  app instead of google developer app
Refer https://github.com/publiclab/plots2/issues/2856
it is simple

SidharthBansal commented:
Refer https://github.com/publiclab/plots2/blob/master/doc/LOGIN_SYSTEMS.md

SidharthBansal commented:
I will be there to guide you. I am not that busy so don't worry
Once #7113 for login and a pr for sign up redirection is complete, then we can go to pr for tests of Mk and pr for tests for SWB login.

Tests for SWB & MK contains 15 points separately if you need them. Else for investigation you will get 5 points apart from redirection issues. 

We will be fair in giving you points, so don't worry.

SidharthBansal commented:
Please get points for installation of MK, SWB and plots2 asap. :-)

Uzay-G commented:
Oh yeah I just took care of that :+1:!

SidharthBansal commented:
In case of any confusion then please feel free to ask( in points/ login systems of MK or SWB or PLots2 etc.)

SidharthBansal commented:
@Uzay-G wherever you find any login/authentication/sign up related issues/pr then kindly add LOGIN & SIGNUP Milestone. If you are unable to do so then kindly ping me

Uzay-G commented:
Hey! @jywarren and @SidharthBansal can we merge this so I can open new pull requests for #5600 and #3367. Thanks!

SidharthBansal commented:
Hi, I am going for sleep right now. If you are fine with reviewing it on
Sunday then please do other tasks in the meantime.
I will review it soon and let you know. :-)

On Wed, 8 Jan 2020, 10:28 pm Uzay-G, <notifications@github.com> wrote:

> Hey! @jywarren <https://github.com/jywarren> and @SidharthBansal
> <https://github.com/SidharthBansal> can we merge this so I can open new
> pull requests for #5600 <https://github.com/publiclab/plots2/issues/5600>
> and #3367 <https://github.com/publiclab/plots2/issues/3367>. Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/plots2/pull/7113?email_source=notifications&email_token=AFAAEQ4RQVRKPYLBGLQEHCTQ4YA2RA5CNFSM4KC3VFC2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEINHXWQ#issuecomment-572160986>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFAAEQ4OU76QTZV6FX4DTY3Q4YA2RANCNFSM4KC3VFCQ>
> .
>

Uzay-G commented:
Hey @SidharthBansal I modified the `sessions_controller` so that when a user signs up for the first time, they are not redirected to the previous page but to the dashboard. This modification was discussed in the issue for this pull request.

Uzay-G commented:
I will also add a test to make sure that when a user signs up through oauth for the first time, they are redirected to the dashboard and **not** the page they were on before signing up.

SidharthBansal commented on app/controllers/user_sessions_controller.rb:
```diff
@@ -77,15 +77,8 @@ def handle_social_login_flow(auth)
             return_to = session[:openid_return_to]
             session[:openid_return_to] = nil
             redirect_to return_to + hash_params
-          elsif params[:return_to] && params[:return_to].split('/')[0..3] == ["", "subscribe", "multiple", "tag"]
```
why have you removed it? This was a feature I built. Any special reasons?

Uzay-G replied:
I removed it because in the issue and when we chatted about oauth, you said when users sign up they should be redirected to the dashboard. This is the code that handles that part:
https://github.com/publiclab/plots2/blob/c4f849d0c9f3a5d54ad9e0ebd6b433ab42dc47ae/app/controllers/user_sessions_controller.rb#L66-L82

So as you can see, if a user without an account goes to `/subscribe/multiple/tag/water` and then they are prompted to sign up, when they signup they will be redirected to the subscriptions instead of the dashboard.

SidharthBansal replied:
I think this part should be discussed separately. Is this needed in this pr?
Title says LOGIN. Isn't it?
We intentionally created this feature due to some reason. Can you please open up an issue for discussion and remove these changes from here if they don't interfere with the OAuth changes?

Uzay-G replied:
Yeah no problem

SidharthBansal replied:
Sure

Uzay-G commented:
I will open up a new issue for signup redirection

SidharthBansal commented:
can I review this pr now?

SidharthBansal commented on test/functional/user_sessions_controller_test.rb:
```diff
@@ -266,4 +266,12 @@ class UserSessionsControllerTest < ActionController::TestCase
     assert_equal "Successfully logged out.",  flash[:notice]
   end
+  
+  test "redirects tto previous page when logging in through oauth" do
```
to typo

SidharthBansal commented:
@IshaGupta18 kindly review this.
LGTM from my side

SidharthBansal commented:
@ananyaarun kindly review this

SidharthBansal commented:
5 points to the PR and 3 points to test. So 8 points.

Uzay-G commented:
Thanks for your great help with this issue @SidharthBansal! I will claim the task on the dashboard

SidharthBansal commented:
Marks given
